{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# OncoBLADE DEMO\n",
    "\n",
    "\n",
    "In this notebook, a demo of OncoBLADE is provided.\n",
    "\n",
    "In this demo we will use a subset of the Non Small Cell Lung Carcinoma (NSCLC) samples from the manuscript to show:\n",
    "1) Cell fraction estimation\n",
    "2) Cell-type specific gene expression estimation\n",
    "3) Application of 2\n",
    "\n",
    "### Move to the right directory and load necessary modules\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, os\n",
    "os.chdir('..') ## Set directory to one up\n",
    "from Deconvolution.OncoBLADE import Framework_Iterative\n",
    "from Deconvolution.OncoBLADE import Purify_AllGenes\n",
    "import numpy as np\n",
    "from numpy import transpose as t\n",
    "import itertools\n",
    "import pickle\n",
    "from scipy.optimize import nnls\n",
    "from sklearn.svm import SVR\n",
    "from sklearn.svm import NuSVR\n",
    "from sklearn.metrics import mean_squared_error as mse\n",
    "import pandas as pd\n",
    "\n",
    "# modules for visualization\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creation of a single cell RNAseq signature\n",
    "\n",
    "OncoBLADE uses a single cell RNAseq (scRNAseq) signature for its bulk RNA deconvolution.\n",
    "Ideally the scRNAseq is from the same type of tumor as the bulk RNAseq, as is the case here with NSCLC.\n",
    "\n",
    "To create the signature we use the phenotyped scRNAseq to calculate a mean (Mu) and expected gene expression variability (Omega) per gene per celltype. The expected gene expression variability was estimated by fitting a LOWESS curve to the mean-variance trend measured in the scRNAseq data using the fitTrendVar function implemented in scran.\n",
    "\n",
    "For this demo we use the same signature as was used in the NSCLC in silico experiments of the manuscript."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        Cancer_cell  Lung_endothelial_cell  Fibroblast  Macrophage  \\\n",
      "A1BG       0.022682               0.072445    0.147478    0.065907   \n",
      "A2M        0.043105               2.459644    1.926918    0.939422   \n",
      "A4GALT     0.201694               0.187970    0.162045    0.005898   \n",
      "AAAS       0.098043               0.058089    0.082713    0.055850   \n",
      "AACS       0.097181               0.028963    0.043312    0.025011   \n",
      "\n",
      "        Plasma_cell  Lung_epithelial_cell  CD4_proliferating_T_cell  Monocyte  \\\n",
      "A1BG       0.112361              0.056572                  0.066149  0.116034   \n",
      "A2M        0.051905              0.045449                  0.046703  0.150999   \n",
      "A4GALT     0.006781              0.104433                  0.007211  0.004006   \n",
      "AAAS       0.052889              0.086882                  0.102814  0.027392   \n",
      "AACS       0.026961              0.053326                  0.032932  0.013841   \n",
      "\n",
      "        Mast_cell        DC  CD8_effector_T_cell    B_cell  Neutrophil  \\\n",
      "A1BG     0.191235  0.154975             0.118646  0.142474    0.000000   \n",
      "A2M      0.036530  0.338606             0.046928  0.040528    0.026962   \n",
      "A4GALT   0.004332  0.002551             0.001266  0.020844    0.001368   \n",
      "AAAS     0.038696  0.065006             0.034847  0.034258    0.004479   \n",
      "AACS     0.023488  0.027993             0.018962  0.014733    0.002787   \n",
      "\n",
      "        Alveolar_cell  CD4_Th17_like_cell  CD4_Treg   NK_cell  \\\n",
      "A1BG         0.019581            0.153735  0.157920  0.062916   \n",
      "A2M          0.029398            0.051364  0.035965  0.045805   \n",
      "A4GALT       0.005708            0.003677  0.006159  0.000000   \n",
      "AAAS         0.075326            0.036957  0.053784  0.041033   \n",
      "AACS         0.054339            0.016715  0.023454  0.031417   \n",
      "\n",
      "        CD4_naive_T_cell  CD8_exhausted_T_cell  \n",
      "A1BG            0.134535              0.174374  \n",
      "A2M             0.023281              0.045238  \n",
      "A4GALT          0.000491              0.000000  \n",
      "AAAS            0.026553              0.050120  \n",
      "AACS            0.023642              0.021126  \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cancer_cell</th>\n",
       "      <th>Lung_endothelial_cell</th>\n",
       "      <th>Fibroblast</th>\n",
       "      <th>Macrophage</th>\n",
       "      <th>Plasma_cell</th>\n",
       "      <th>Lung_epithelial_cell</th>\n",
       "      <th>CD4_proliferating_T_cell</th>\n",
       "      <th>Monocyte</th>\n",
       "      <th>Mast_cell</th>\n",
       "      <th>DC</th>\n",
       "      <th>CD8_effector_T_cell</th>\n",
       "      <th>B_cell</th>\n",
       "      <th>Neutrophil</th>\n",
       "      <th>Alveolar_cell</th>\n",
       "      <th>CD4_Th17_like_cell</th>\n",
       "      <th>CD4_Treg</th>\n",
       "      <th>NK_cell</th>\n",
       "      <th>CD4_naive_T_cell</th>\n",
       "      <th>CD8_exhausted_T_cell</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.036189</td>\n",
       "      <td>0.098136</td>\n",
       "      <td>0.181896</td>\n",
       "      <td>0.080758</td>\n",
       "      <td>0.155971</td>\n",
       "      <td>0.071124</td>\n",
       "      <td>0.088225</td>\n",
       "      <td>0.161703</td>\n",
       "      <td>0.270806</td>\n",
       "      <td>0.162115</td>\n",
       "      <td>0.208026</td>\n",
       "      <td>0.197206</td>\n",
       "      <td>0.010000</td>\n",
       "      <td>0.035627</td>\n",
       "      <td>0.230141</td>\n",
       "      <td>0.223709</td>\n",
       "      <td>0.098819</td>\n",
       "      <td>0.184399</td>\n",
       "      <td>0.219604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.059641</td>\n",
       "      <td>1.187655</td>\n",
       "      <td>1.226067</td>\n",
       "      <td>0.803350</td>\n",
       "      <td>0.078141</td>\n",
       "      <td>0.059110</td>\n",
       "      <td>0.065231</td>\n",
       "      <td>0.205610</td>\n",
       "      <td>0.061031</td>\n",
       "      <td>0.324161</td>\n",
       "      <td>0.088831</td>\n",
       "      <td>0.064017</td>\n",
       "      <td>0.076573</td>\n",
       "      <td>0.048453</td>\n",
       "      <td>0.084862</td>\n",
       "      <td>0.059597</td>\n",
       "      <td>0.074665</td>\n",
       "      <td>0.040504</td>\n",
       "      <td>0.065863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>0.236096</td>\n",
       "      <td>0.233397</td>\n",
       "      <td>0.198340</td>\n",
       "      <td>0.016346</td>\n",
       "      <td>0.018961</td>\n",
       "      <td>0.122678</td>\n",
       "      <td>0.018528</td>\n",
       "      <td>0.015260</td>\n",
       "      <td>0.016052</td>\n",
       "      <td>0.012545</td>\n",
       "      <td>0.012126</td>\n",
       "      <td>0.037782</td>\n",
       "      <td>0.013378</td>\n",
       "      <td>0.017474</td>\n",
       "      <td>0.015360</td>\n",
       "      <td>0.018494</td>\n",
       "      <td>0.010000</td>\n",
       "      <td>0.010643</td>\n",
       "      <td>0.010000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>0.121755</td>\n",
       "      <td>0.080672</td>\n",
       "      <td>0.107298</td>\n",
       "      <td>0.069993</td>\n",
       "      <td>0.079423</td>\n",
       "      <td>0.103837</td>\n",
       "      <td>0.131474</td>\n",
       "      <td>0.045965</td>\n",
       "      <td>0.064057</td>\n",
       "      <td>0.074844</td>\n",
       "      <td>0.068537</td>\n",
       "      <td>0.055660</td>\n",
       "      <td>0.021060</td>\n",
       "      <td>0.108113</td>\n",
       "      <td>0.063866</td>\n",
       "      <td>0.084168</td>\n",
       "      <td>0.067927</td>\n",
       "      <td>0.044791</td>\n",
       "      <td>0.071892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACS</th>\n",
       "      <td>0.120793</td>\n",
       "      <td>0.045239</td>\n",
       "      <td>0.060974</td>\n",
       "      <td>0.036898</td>\n",
       "      <td>0.045523</td>\n",
       "      <td>0.067618</td>\n",
       "      <td>0.048946</td>\n",
       "      <td>0.028173</td>\n",
       "      <td>0.042812</td>\n",
       "      <td>0.037925</td>\n",
       "      <td>0.041852</td>\n",
       "      <td>0.029637</td>\n",
       "      <td>0.016882</td>\n",
       "      <td>0.080934</td>\n",
       "      <td>0.034362</td>\n",
       "      <td>0.042344</td>\n",
       "      <td>0.054352</td>\n",
       "      <td>0.040977</td>\n",
       "      <td>0.036088</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Cancer_cell  Lung_endothelial_cell  Fibroblast  Macrophage  \\\n",
       "A1BG       0.036189               0.098136    0.181896    0.080758   \n",
       "A2M        0.059641               1.187655    1.226067    0.803350   \n",
       "A4GALT     0.236096               0.233397    0.198340    0.016346   \n",
       "AAAS       0.121755               0.080672    0.107298    0.069993   \n",
       "AACS       0.120793               0.045239    0.060974    0.036898   \n",
       "\n",
       "        Plasma_cell  Lung_epithelial_cell  CD4_proliferating_T_cell  Monocyte  \\\n",
       "A1BG       0.155971              0.071124                  0.088225  0.161703   \n",
       "A2M        0.078141              0.059110                  0.065231  0.205610   \n",
       "A4GALT     0.018961              0.122678                  0.018528  0.015260   \n",
       "AAAS       0.079423              0.103837                  0.131474  0.045965   \n",
       "AACS       0.045523              0.067618                  0.048946  0.028173   \n",
       "\n",
       "        Mast_cell        DC  CD8_effector_T_cell    B_cell  Neutrophil  \\\n",
       "A1BG     0.270806  0.162115             0.208026  0.197206    0.010000   \n",
       "A2M      0.061031  0.324161             0.088831  0.064017    0.076573   \n",
       "A4GALT   0.016052  0.012545             0.012126  0.037782    0.013378   \n",
       "AAAS     0.064057  0.074844             0.068537  0.055660    0.021060   \n",
       "AACS     0.042812  0.037925             0.041852  0.029637    0.016882   \n",
       "\n",
       "        Alveolar_cell  CD4_Th17_like_cell  CD4_Treg   NK_cell  \\\n",
       "A1BG         0.035627            0.230141  0.223709  0.098819   \n",
       "A2M          0.048453            0.084862  0.059597  0.074665   \n",
       "A4GALT       0.017474            0.015360  0.018494  0.010000   \n",
       "AAAS         0.108113            0.063866  0.084168  0.067927   \n",
       "AACS         0.080934            0.034362  0.042344  0.054352   \n",
       "\n",
       "        CD4_naive_T_cell  CD8_exhausted_T_cell  \n",
       "A1BG            0.184399              0.219604  \n",
       "A2M             0.040504              0.065863  \n",
       "A4GALT          0.010643              0.010000  \n",
       "AAAS            0.044791              0.071892  \n",
       "AACS            0.040977              0.036088  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load NSCLC signature\n",
    "with open('data/NewSignature_matrices_final_celltype_k1000.pickle', \"rb\") as infile:\n",
    "    signature = pickle.load(infile)\n",
    "\n",
    "## Extract Mu & Omega (nGenes x nCelltypes)\n",
    "Mu = pd.DataFrame(signature['Mu'], index = signature['Genelist'], columns = signature['celltype_list'])\n",
    "Omega = pd.DataFrame(signature['Omega'], index = signature['Genelist'], columns = signature['celltype_list'])\n",
    "\n",
    "## Add pseudocount to Omega as 0 will otherwise give computational problems\n",
    "Omega = Omega + 0.01\n",
    "## Print first 5 rows\n",
    "print(Mu.head())\n",
    "Omega.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preparation of Bulk RNAseq and expected tumor fraction\n",
    "The main innovation of OncoBLADE with respect to other RNA deconvolution methods, is that it allows the user to inform it with cell fraction estimates. OncoBLADE can for example be informed by DNA-derived malignant fraction estimates, making the rest of the deconvolution task significantly easier.\n",
    "Besides malignant cell fraction estimates, you can also include information on other cell types if you have it. For example if you have Immunohistochemistry stainings on your samples.\n",
    "\n",
    "Below we show how to prepare your prior fraction expectation and the bulk RNAseq before performing the deconvolution.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-91-6840</th>\n",
       "      <th>TCGA-78-8655</th>\n",
       "      <th>TCGA-NJ-A7XG</th>\n",
       "      <th>TCGA-38-4625</th>\n",
       "      <th>TCGA-38-4627</th>\n",
       "      <th>TCGA-85-A4QQ</th>\n",
       "      <th>TCGA-56-8628</th>\n",
       "      <th>TCGA-63-A5M9</th>\n",
       "      <th>TCGA-77-8128</th>\n",
       "      <th>TCGA-34-5231</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "      <td>1.643189</td>\n",
       "      <td>1.005787</td>\n",
       "      <td>0.986710</td>\n",
       "      <td>1.015124</td>\n",
       "      <td>1.559100</td>\n",
       "      <td>0.308683</td>\n",
       "      <td>0.829310</td>\n",
       "      <td>0.904077</td>\n",
       "      <td>0.727355</td>\n",
       "      <td>0.653832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNMD</th>\n",
       "      <td>0.001019</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000437</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000496</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>0.478754</td>\n",
       "      <td>0.269087</td>\n",
       "      <td>0.275330</td>\n",
       "      <td>0.673312</td>\n",
       "      <td>0.293269</td>\n",
       "      <td>0.452470</td>\n",
       "      <td>0.283463</td>\n",
       "      <td>0.516275</td>\n",
       "      <td>0.456260</td>\n",
       "      <td>0.378602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>0.185521</td>\n",
       "      <td>0.295398</td>\n",
       "      <td>0.194395</td>\n",
       "      <td>0.117442</td>\n",
       "      <td>0.105425</td>\n",
       "      <td>0.106066</td>\n",
       "      <td>0.144088</td>\n",
       "      <td>0.172375</td>\n",
       "      <td>0.179041</td>\n",
       "      <td>0.200384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>0.148485</td>\n",
       "      <td>0.084015</td>\n",
       "      <td>0.056444</td>\n",
       "      <td>0.227574</td>\n",
       "      <td>0.054897</td>\n",
       "      <td>0.111430</td>\n",
       "      <td>0.073408</td>\n",
       "      <td>0.320247</td>\n",
       "      <td>0.160673</td>\n",
       "      <td>0.174342</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          TCGA-91-6840  TCGA-78-8655  TCGA-NJ-A7XG  TCGA-38-4625  \\\n",
       "symbol                                                             \n",
       "TSPAN6        1.643189      1.005787      0.986710      1.015124   \n",
       "TNMD          0.001019      0.000000      0.000000      0.000000   \n",
       "DPM1          0.478754      0.269087      0.275330      0.673312   \n",
       "SCYL3         0.185521      0.295398      0.194395      0.117442   \n",
       "C1orf112      0.148485      0.084015      0.056444      0.227574   \n",
       "\n",
       "          TCGA-38-4627  TCGA-85-A4QQ  TCGA-56-8628  TCGA-63-A5M9  \\\n",
       "symbol                                                             \n",
       "TSPAN6        1.559100      0.308683      0.829310      0.904077   \n",
       "TNMD          0.000437      0.000000      0.000496      0.000000   \n",
       "DPM1          0.293269      0.452470      0.283463      0.516275   \n",
       "SCYL3         0.105425      0.106066      0.144088      0.172375   \n",
       "C1orf112      0.054897      0.111430      0.073408      0.320247   \n",
       "\n",
       "          TCGA-77-8128  TCGA-34-5231  \n",
       "symbol                                \n",
       "TSPAN6        0.727355      0.653832  \n",
       "TNMD          0.000000      0.000099  \n",
       "DPM1          0.456260      0.378602  \n",
       "SCYL3         0.179041      0.200384  \n",
       "C1orf112      0.160673      0.174342  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Load in bulk RNAseq, first 5 patients are LUAD, last 5 are LUSC\n",
    "bulk = pd.read_csv('data/Transcriptome_matrix_subset.txt', sep = '\\t')\n",
    "\n",
    "## Normalize bulk RNAseq to the same scale as the RNAseq, we use counts per 10k.\n",
    "## Do not Log normalize, this is done within OncoBLADE\n",
    "samples = [s for s in bulk.columns if 'TCGA' in s]\n",
    "bulk[samples] = bulk[samples].apply(lambda x: x / sum(x) * 10000)\n",
    "\n",
    "# set index to gene symbols and put normalized bulk in Y\n",
    "Y =  bulk.set_index('symbol')\n",
    "Y.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[0.55  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan\n",
      "   nan  nan  nan  nan  nan]\n",
      " [0.53  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan\n",
      "   nan  nan  nan  nan  nan]\n",
      " [0.43  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan\n",
      "   nan  nan  nan  nan  nan]\n",
      " [0.2   nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan\n",
      "   nan  nan  nan  nan  nan]\n",
      " [0.2   nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan\n",
      "   nan  nan  nan  nan  nan]\n",
      " [0.28  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan\n",
      "   nan  nan  nan  nan  nan]\n",
      " [0.41  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan\n",
      "   nan  nan  nan  nan  nan]\n",
      " [0.64  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan\n",
      "   nan  nan  nan  nan  nan]\n",
      " [0.43  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan\n",
      "   nan  nan  nan  nan  nan]\n",
      " [0.44  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan\n",
      "   nan  nan  nan  nan  nan]]\n"
     ]
    }
   ],
   "source": [
    "## Prepare prior expectation\n",
    "# Load ACE tumor fraction estimates\n",
    "purities = pd.read_csv('data/ACE_Tumor_purities_squaremodel.tsv', sep = \"\\t\")\n",
    "\n",
    "# obtain list of expected tumor purities for the 10 samples\n",
    "expected_tumor_purity =pd.merge(pd.DataFrame({'sample': samples }),purities, how = 'left').ACE.tolist()\n",
    "\n",
    "# Intialize Expectation (Nsample x Ncell with None for non-tumor celltypes)\n",
    "Expectation = np.zeros((len(samples), len(signature['celltype_list']))) + np.nan\n",
    "\n",
    "# iterate over samples\n",
    "for i in range(len(samples)):\n",
    "    # iterate over celltypes\n",
    "    for j,celltype in enumerate(signature['celltype_list']):\n",
    "        if celltype in ['Cancer_cell', 'Tumor cell']:\n",
    "            # fetch true tumor purity and add to array\n",
    "            Expectation[i,j] = expected_tumor_purity[i]\n",
    "        else:\n",
    "            pass\n",
    "\n",
    "print(Expectation)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Subset signature & bulk on selected genes for deconvolution\n",
    "\n",
    "OncoBLADE like other RNA deconvolution tools uses a selection of genes for its deconvolution due to the otherwise enormous complexity.\n",
    "\n",
    "For this we want genes discrimating the cell types which are to be deconvoluted. We use AutoGeneS for this and use the genes it selected on our scRNAseq here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-91-6840</th>\n",
       "      <th>TCGA-78-8655</th>\n",
       "      <th>TCGA-NJ-A7XG</th>\n",
       "      <th>TCGA-38-4625</th>\n",
       "      <th>TCGA-38-4627</th>\n",
       "      <th>TCGA-85-A4QQ</th>\n",
       "      <th>TCGA-56-8628</th>\n",
       "      <th>TCGA-63-A5M9</th>\n",
       "      <th>TCGA-77-8128</th>\n",
       "      <th>TCGA-34-5231</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>10.682426</td>\n",
       "      <td>10.446552</td>\n",
       "      <td>1.925969</td>\n",
       "      <td>5.821721</td>\n",
       "      <td>23.110129</td>\n",
       "      <td>0.628959</td>\n",
       "      <td>13.675426</td>\n",
       "      <td>3.468265</td>\n",
       "      <td>2.097612</td>\n",
       "      <td>3.076826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCA1</th>\n",
       "      <td>0.746502</td>\n",
       "      <td>0.447581</td>\n",
       "      <td>0.163208</td>\n",
       "      <td>0.999605</td>\n",
       "      <td>0.387482</td>\n",
       "      <td>0.533274</td>\n",
       "      <td>0.677782</td>\n",
       "      <td>0.248778</td>\n",
       "      <td>0.751002</td>\n",
       "      <td>0.190345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCA3</th>\n",
       "      <td>0.451231</td>\n",
       "      <td>1.060800</td>\n",
       "      <td>3.984150</td>\n",
       "      <td>1.075697</td>\n",
       "      <td>0.364474</td>\n",
       "      <td>0.055542</td>\n",
       "      <td>4.889802</td>\n",
       "      <td>0.343559</td>\n",
       "      <td>0.237948</td>\n",
       "      <td>0.234378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC022182.3</th>\n",
       "      <td>0.001019</td>\n",
       "      <td>0.000598</td>\n",
       "      <td>0.000383</td>\n",
       "      <td>0.000801</td>\n",
       "      <td>0.001747</td>\n",
       "      <td>0.000173</td>\n",
       "      <td>0.002480</td>\n",
       "      <td>0.001531</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC133644.2</th>\n",
       "      <td>0.008155</td>\n",
       "      <td>0.011660</td>\n",
       "      <td>0.001913</td>\n",
       "      <td>0.007209</td>\n",
       "      <td>0.016455</td>\n",
       "      <td>0.002768</td>\n",
       "      <td>0.003968</td>\n",
       "      <td>0.008168</td>\n",
       "      <td>0.008023</td>\n",
       "      <td>0.004373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZMIZ1</th>\n",
       "      <td>2.380857</td>\n",
       "      <td>0.627272</td>\n",
       "      <td>0.478144</td>\n",
       "      <td>0.497400</td>\n",
       "      <td>0.909657</td>\n",
       "      <td>1.252553</td>\n",
       "      <td>1.273724</td>\n",
       "      <td>0.392907</td>\n",
       "      <td>1.153000</td>\n",
       "      <td>1.295239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZMYND10</th>\n",
       "      <td>0.090382</td>\n",
       "      <td>0.008073</td>\n",
       "      <td>0.064288</td>\n",
       "      <td>0.010513</td>\n",
       "      <td>0.078923</td>\n",
       "      <td>0.079593</td>\n",
       "      <td>0.048360</td>\n",
       "      <td>0.004594</td>\n",
       "      <td>0.007390</td>\n",
       "      <td>0.048804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF276</th>\n",
       "      <td>0.193336</td>\n",
       "      <td>0.240683</td>\n",
       "      <td>0.235915</td>\n",
       "      <td>0.076893</td>\n",
       "      <td>0.039753</td>\n",
       "      <td>0.243278</td>\n",
       "      <td>0.412919</td>\n",
       "      <td>0.118093</td>\n",
       "      <td>0.164051</td>\n",
       "      <td>0.218176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF486</th>\n",
       "      <td>0.958187</td>\n",
       "      <td>0.108233</td>\n",
       "      <td>0.005357</td>\n",
       "      <td>0.113036</td>\n",
       "      <td>0.223665</td>\n",
       "      <td>0.002076</td>\n",
       "      <td>0.087296</td>\n",
       "      <td>0.172716</td>\n",
       "      <td>0.010346</td>\n",
       "      <td>0.007952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF683</th>\n",
       "      <td>0.060481</td>\n",
       "      <td>0.077736</td>\n",
       "      <td>0.009184</td>\n",
       "      <td>0.014918</td>\n",
       "      <td>0.009756</td>\n",
       "      <td>0.024224</td>\n",
       "      <td>0.036456</td>\n",
       "      <td>0.018888</td>\n",
       "      <td>0.009079</td>\n",
       "      <td>0.024949</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>548 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            TCGA-91-6840  TCGA-78-8655  TCGA-NJ-A7XG  TCGA-38-4625  \\\n",
       "symbol                                                               \n",
       "A2M            10.682426     10.446552      1.925969      5.821721   \n",
       "ABCA1           0.746502      0.447581      0.163208      0.999605   \n",
       "ABCA3           0.451231      1.060800      3.984150      1.075697   \n",
       "AC022182.3      0.001019      0.000598      0.000383      0.000801   \n",
       "AC133644.2      0.008155      0.011660      0.001913      0.007209   \n",
       "...                  ...           ...           ...           ...   \n",
       "ZMIZ1           2.380857      0.627272      0.478144      0.497400   \n",
       "ZMYND10         0.090382      0.008073      0.064288      0.010513   \n",
       "ZNF276          0.193336      0.240683      0.235915      0.076893   \n",
       "ZNF486          0.958187      0.108233      0.005357      0.113036   \n",
       "ZNF683          0.060481      0.077736      0.009184      0.014918   \n",
       "\n",
       "            TCGA-38-4627  TCGA-85-A4QQ  TCGA-56-8628  TCGA-63-A5M9  \\\n",
       "symbol                                                               \n",
       "A2M            23.110129      0.628959     13.675426      3.468265   \n",
       "ABCA1           0.387482      0.533274      0.677782      0.248778   \n",
       "ABCA3           0.364474      0.055542      4.889802      0.343559   \n",
       "AC022182.3      0.001747      0.000173      0.002480      0.001531   \n",
       "AC133644.2      0.016455      0.002768      0.003968      0.008168   \n",
       "...                  ...           ...           ...           ...   \n",
       "ZMIZ1           0.909657      1.252553      1.273724      0.392907   \n",
       "ZMYND10         0.078923      0.079593      0.048360      0.004594   \n",
       "ZNF276          0.039753      0.243278      0.412919      0.118093   \n",
       "ZNF486          0.223665      0.002076      0.087296      0.172716   \n",
       "ZNF683          0.009756      0.024224      0.036456      0.018888   \n",
       "\n",
       "            TCGA-77-8128  TCGA-34-5231  \n",
       "symbol                                  \n",
       "A2M             2.097612      3.076826  \n",
       "ABCA1           0.751002      0.190345  \n",
       "ABCA3           0.237948      0.234378  \n",
       "AC022182.3      0.000000      0.000795  \n",
       "AC133644.2      0.008023      0.004373  \n",
       "...                  ...           ...  \n",
       "ZMIZ1           1.153000      1.295239  \n",
       "ZMYND10         0.007390      0.048804  \n",
       "ZNF276          0.164051      0.218176  \n",
       "ZNF486          0.010346      0.007952  \n",
       "ZNF683          0.009079      0.024949  \n",
       "\n",
       "[548 rows x 10 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Load autogenes which was saved in the signature\n",
    "AutoGeneS = signature['AutoGeneS']['Unnamed: 0'].tolist()\n",
    "\n",
    "## Find common AutoGeneS with the bulk data\n",
    "common_AutoGeneS = [value for value in AutoGeneS if value in Y.index.to_list()]\n",
    "\n",
    "## Subset Mu, Omega & bulk on AutoGeneS\n",
    "Mu_AutoGeneS = Mu.loc[common_AutoGeneS,]\n",
    "Omega_AutoGeneS = Omega.loc[common_AutoGeneS,]\n",
    "Y_AutoGeneS = Y.loc[common_AutoGeneS,]\n",
    "Y_AutoGeneS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setting the parameters for OncoBLADE\n",
    "Before running OncoBLADE there is one more thing to do, setting the (hyper)parameters.\n",
    "The key parameters used in OncoBLADE are:\n",
    "- Hyperparameters (`hyperpars`): `Alpha`, `Alpha0`, `Kappa0` and `SigmaY`.\n",
    "- `Nrep`: Number of repeats for evaluating each parameter configuration.\n",
    "- `Nrepfinal`: Number of repeated optimizations for the final parameter set.\n",
    "- `Njob`: Number of parallel jobs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "pars = {\n",
    "    'Alpha': 1,\n",
    "    'Alpha0': 1000,\n",
    "    'Kappa0': 1,\n",
    "    'SY': 1,\n",
    "    'Nrep': 3, ## small for demo purposes, for real application >10 is advised\n",
    "    'Nrepfinal': 100, ## small for demo purposes, for real application >1000 is advised\n",
    "    'Njob': 4 ## Number of parallel jobs.\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1) Cell fraction estimation by OncoBLADE\n",
    "Now we are ready to run the first step of OncoBLADE, the cell fraction estimation. Here it is applied to 10 NSCLC samples.\n",
    "\n",
    "OncoBLADE produce several outcomes:\n",
    "- `final_obj`: final BLADE object with optimized variational parameters\n",
    "- `best_obj`: BLADE object trained with the best parameter set found by the Empirical Bayes framework. Empirical Bayes framework is applied after selecting a subset of samples (5 samples; indicated by `Ind_sample` below), and thus the outcome contains only 5 samples. If `Nsample` <= 5, `final_obj` is identical to `best_obj`.\n",
    "- `best_set`: Best parameter set defined by Empirical Bayes framework.\n",
    "- `outs`: Outcome of BLADE for every possible combination of hyperparameters, used in the Empirical Bayes framework. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all of 548 genes are used for optimization.\n",
      "Initialization with Support vector regression\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=4)]: Using backend LokyBackend with 4 concurrent workers.\n",
      "[Parallel(n_jobs=4)]: Done   5 out of  10 | elapsed:    2.4s remaining:    2.4s\n",
      "[Parallel(n_jobs=4)]: Done   7 out of  10 | elapsed:    2.7s remaining:    1.2s\n",
      "[Parallel(n_jobs=4)]: Done  10 out of  10 | elapsed:    3.1s finished\n",
      "[Parallel(n_jobs=4)]: Using backend LokyBackend with 4 concurrent workers.\n",
      "/net/beegfs/cfg/tgac/m.steketee/OncoBLADE/Deconvolution/OncoBLADE.py:871: Warning: No optimization is not done yet\n",
      "  obj.Check_health()\n",
      "/net/beegfs/cfg/tgac/m.steketee/OncoBLADE/Deconvolution/OncoBLADE.py:871: Warning: No optimization is not done yet\n",
      "  obj.Check_health()\n",
      "/net/beegfs/cfg/tgac/m.steketee/OncoBLADE/Deconvolution/OncoBLADE.py:871: Warning: No optimization is not done yet\n",
      "  obj.Check_health()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No feature filtering is done (fsel = 0)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=4)]: Done   3 out of   3 | elapsed: 13.8min remaining:    0.0s\n",
      "[Parallel(n_jobs=4)]: Done   3 out of   3 | elapsed: 13.8min finished\n"
     ]
    }
   ],
   "source": [
    "final_obj, best_obj, best_set, outs = Framework_Iterative(\n",
    "            Mu_AutoGeneS.to_numpy(), Omega_AutoGeneS.to_numpy(),Y_AutoGeneS.to_numpy(),\n",
    "            Alpha=pars['Alpha'], Alpha0=pars['Alpha0'], \n",
    "            Kappa0=pars['Kappa0'], sY = pars['SY'],\n",
    "            Nrep= pars['Nrep'], Njob= pars['Njob'], IterMax= pars['Nrepfinal'], Expectation = Expectation)\n",
    "\n",
    "## Save BLADE output in one dictionary\n",
    "BLADE_output = {\n",
    "    'final_obj': final_obj,\n",
    "    'best_obj': best_obj,\n",
    "    'best_set': best_set,\n",
    "    'outs' : outs,\n",
    "    'genes' : common_AutoGeneS\n",
    "     }\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2) Cell type specific gene expression estimation by OncoBLADE\n",
    "After step 1, we can fix the estimated cell fractions which allows us to estimate the cell type specific gene expression in parallel per gene. Here we estimate the cell type specific gene expression levels for the top 1819 highly variable genes in the scRNAseq data.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=4)]: Using backend LokyBackend with 4 concurrent workers.\n",
      "/net/beegfs/cfg/tgac/m.steketee/OncoBLADE/Deconvolution/OncoBLADE.py:974: Warning: No optimization is not done yet\n",
      "  obj.Check_health()\n",
      "/net/beegfs/cfg/tgac/m.steketee/OncoBLADE/Deconvolution/OncoBLADE.py:974: Warning: No optimization is not done yet\n",
      "  obj.Check_health()\n",
      "/net/beegfs/cfg/tgac/m.steketee/OncoBLADE/Deconvolution/OncoBLADE.py:974: Warning: No optimization is not done yet\n",
      "  obj.Check_health()\n",
      "[Parallel(n_jobs=4)]: Done   5 tasks      | elapsed:    1.6s\n",
      "[Parallel(n_jobs=4)]: Done  10 tasks      | elapsed:    2.4s\n",
      "[Parallel(n_jobs=4)]: Done  17 tasks      | elapsed:    3.9s\n",
      "[Parallel(n_jobs=4)]: Done  24 tasks      | elapsed:    5.1s\n",
      "[Parallel(n_jobs=4)]: Done  33 tasks      | elapsed:    7.4s\n",
      "[Parallel(n_jobs=4)]: Done  42 tasks      | elapsed:   10.5s\n",
      "[Parallel(n_jobs=4)]: Done  53 tasks      | elapsed:   13.2s\n",
      "/net/beegfs/cfg/tgac/m.steketee/OncoBLADE/Deconvolution/OncoBLADE.py:974: Warning: No optimization is not done yet\n",
      "  obj.Check_health()\n",
      "[Parallel(n_jobs=4)]: Done  64 tasks      | elapsed:   15.5s\n",
      "[Parallel(n_jobs=4)]: Done  77 tasks      | elapsed:   18.4s\n",
      "[Parallel(n_jobs=4)]: Done  90 tasks      | elapsed:   21.1s\n",
      "[Parallel(n_jobs=4)]: Done 105 tasks      | elapsed:   23.8s\n",
      "[Parallel(n_jobs=4)]: Done 120 tasks      | elapsed:   27.5s\n",
      "/net/beegfs/cfg/tgac/m.steketee/OncoBLADE/Deconvolution/OncoBLADE.py:974: Warning: No optimization is not done yet\n",
      "  obj.Check_health()\n",
      "[Parallel(n_jobs=4)]: Done 137 tasks      | elapsed:   30.2s\n",
      "[Parallel(n_jobs=4)]: Done 154 tasks      | elapsed:   33.5s\n",
      "[Parallel(n_jobs=4)]: Done 173 tasks      | elapsed:   36.6s\n",
      "[Parallel(n_jobs=4)]: Done 192 tasks      | elapsed:   40.7s\n",
      "[Parallel(n_jobs=4)]: Done 213 tasks      | elapsed:   43.7s\n",
      "[Parallel(n_jobs=4)]: Done 234 tasks      | elapsed:   46.8s\n",
      "[Parallel(n_jobs=4)]: Done 257 tasks      | elapsed:   50.9s\n",
      "[Parallel(n_jobs=4)]: Done 280 tasks      | elapsed:   55.5s\n",
      "[Parallel(n_jobs=4)]: Done 305 tasks      | elapsed:   59.7s\n",
      "[Parallel(n_jobs=4)]: Done 330 tasks      | elapsed:  1.1min\n",
      "[Parallel(n_jobs=4)]: Done 357 tasks      | elapsed:  1.1min\n",
      "[Parallel(n_jobs=4)]: Done 384 tasks      | elapsed:  1.2min\n",
      "[Parallel(n_jobs=4)]: Done 413 tasks      | elapsed:  1.3min\n",
      "[Parallel(n_jobs=4)]: Done 442 tasks      | elapsed:  1.4min\n",
      "[Parallel(n_jobs=4)]: Done 473 tasks      | elapsed:  1.5min\n",
      "[Parallel(n_jobs=4)]: Done 504 tasks      | elapsed:  1.6min\n",
      "[Parallel(n_jobs=4)]: Done 537 tasks      | elapsed:  1.7min\n",
      "[Parallel(n_jobs=4)]: Done 570 tasks      | elapsed:  1.8min\n",
      "[Parallel(n_jobs=4)]: Done 605 tasks      | elapsed:  1.9min\n",
      "[Parallel(n_jobs=4)]: Done 640 tasks      | elapsed:  2.0min\n",
      "[Parallel(n_jobs=4)]: Done 677 tasks      | elapsed:  2.1min\n",
      "[Parallel(n_jobs=4)]: Done 714 tasks      | elapsed:  2.2min\n",
      "[Parallel(n_jobs=4)]: Done 753 tasks      | elapsed:  2.3min\n",
      "[Parallel(n_jobs=4)]: Done 792 tasks      | elapsed:  2.4min\n",
      "[Parallel(n_jobs=4)]: Done 833 tasks      | elapsed:  2.6min\n",
      "[Parallel(n_jobs=4)]: Done 874 tasks      | elapsed:  2.7min\n",
      "[Parallel(n_jobs=4)]: Done 917 tasks      | elapsed:  2.8min\n",
      "[Parallel(n_jobs=4)]: Done 960 tasks      | elapsed:  2.9min\n",
      "[Parallel(n_jobs=4)]: Done 1005 tasks      | elapsed:  3.0min\n",
      "[Parallel(n_jobs=4)]: Done 1050 tasks      | elapsed:  3.2min\n",
      "[Parallel(n_jobs=4)]: Done 1097 tasks      | elapsed:  3.3min\n",
      "[Parallel(n_jobs=4)]: Done 1144 tasks      | elapsed:  3.4min\n",
      "[Parallel(n_jobs=4)]: Done 1193 tasks      | elapsed:  3.6min\n",
      "[Parallel(n_jobs=4)]: Done 1242 tasks      | elapsed:  3.7min\n",
      "[Parallel(n_jobs=4)]: Done 1293 tasks      | elapsed:  3.8min\n",
      "[Parallel(n_jobs=4)]: Done 1344 tasks      | elapsed:  4.0min\n",
      "[Parallel(n_jobs=4)]: Done 1397 tasks      | elapsed:  4.1min\n",
      "[Parallel(n_jobs=4)]: Done 1450 tasks      | elapsed:  4.3min\n",
      "[Parallel(n_jobs=4)]: Done 1505 tasks      | elapsed:  4.4min\n",
      "[Parallel(n_jobs=4)]: Done 1560 tasks      | elapsed:  4.6min\n",
      "[Parallel(n_jobs=4)]: Done 1617 tasks      | elapsed:  4.7min\n",
      "[Parallel(n_jobs=4)]: Done 1674 tasks      | elapsed:  4.9min\n"
     ]
    }
   ],
   "source": [
    "## Load autogenes which was saved in the signature\n",
    "hvgenes = signature['GeneList_hvg']\n",
    "\n",
    "## Find common hvg with the bulk data\n",
    "common_hvgenes = [value for value in hvgenes if value in Y.index.to_list()]\n",
    "\n",
    "## Subset Mu, Omega & bulk on AutoGeneS\n",
    "Mu_hvgenes = Mu.loc[common_hvgenes,]\n",
    "Omega_hvgenes = Omega.loc[common_hvgenes,]\n",
    "Y_hvgenes = Y.loc[common_hvgenes,]\n",
    "\n",
    "## Estimate cell type specific gene expression\n",
    "final_obj_2, obj_func = Purify_AllGenes(BLADE_output, Mu_hvgenes.to_numpy(), Omega_hvgenes.to_numpy(), Y_hvgenes.to_numpy(), pars['Njob'])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
